src/linear_algebra/projection.lean

```diff
@@ -167,6 +167,77 @@ lemma linear_proj_of_is_compl_idempotent (h : is_compl p q) (x : E) :
     linear_proj_of_is_compl p q h x :=
 linear_proj_of_is_compl_apply_left h _
 
+lemma exists_unique_add_of_is_compl (hc : is_compl p q) (x : E) :
+  ∃ (u : p) (v : q), (u.1 + v.1 = x ∧ ∀ (r : p) (s : q),
+    r.1 + s.1 = x → r = u ∧ s = v) :=
+begin
+  obtain ⟨u, v, h⟩ :=
+    submodule.mem_sup'.1 ((eq_top_iff.2 hc.2).symm ▸ mem_top : x ∈ p ⊔ q),
+  refine ⟨u, v, h, λ r s hrs, _⟩,
+  erw [← h, ← @add_right_cancel_iff _ _ (-s : E) (r + s) (u + v),
+       add_assoc, add_neg_self, add_zero,
+       ← @add_left_cancel_iff _ _ (-u : E) r (u + v + -s),
+       ← add_assoc, ← add_assoc, add_left_neg, zero_add, add_comm,
+       ← sub_eq_add_neg, ← sub_eq_add_neg] at hrs,
+  split; rw [← sub_eq_zero],
+  { suffices : r.1 - u.1 ∈ q,
+    { rw [← coe_eq_coe, coe_sub],
+      exact (submodule.mem_bot _).1 (hc.1 ⟨p.sub_mem r.2 u.2, this⟩) },
+    erw hrs, exact q.sub_mem v.2 s.2 },
+  { suffices : s.1 - v.1 ∈ p,
+    { rw [← coe_eq_coe, coe_sub],
+      exact (submodule.mem_bot _).1 (hc.1 ⟨this, q.sub_mem s.2 v.2⟩) },
+    erw [← neg_mem_iff, neg_sub, ← hrs],
+    exact p.sub_mem r.2 u.2 }
+end .
+
+/-- Given linear maps `φ` and `ψ` from complement submodules, `of_is_compl` is
+the induced linear map over the entire module. -/
+noncomputable def of_is_compl
+  (h : is_compl p q) (φ : p →ₗ[R] F) (ψ : q →ₗ[R] F) : E →ₗ[R] F :=
```

Comment:
I think you can bundle this whole thing into a linear equiv between `prod (p →ₗ[R] F) (q →ₗ[R] F)` and `E →ₗ[R] F`? That would subsume all your lemmas below I think

Reply:
In fact, I think `prod_equiv_of_is_compl` is exactly that statement...

Reply:
I see, something like this? `noncomputable def of_is_compl
  {R : Type*} [comm_ring R] {E : Type*} [add_comm_group E] [module R E]
  {F : Type*} [add_comm_group F] [module R F] {p q : submodule R E}
  (h : is_compl p q) : (p →ₗ[R] F) × (q →ₗ[R] F) ≃ₗ[R] (E →ₗ[R] F) := sorry` using `prod_equiv_of_is_compl` is doable

Reply:
But we can only do this for `comm_ring R` since otherwise `(q →ₗ[R] F)` is not a module over `R`. Is it worthwhile to have `of_is_compl` for general rings?

Reply:
Ah, then I guess just make it an add_equiv

Reply:
Actually, I don't think the method using `prod_equiv_of_is_compl` work even for `comm_ring` since we cannot prove `is_compl (p →ₗ[R] F) (q →ₗ[R] F)` since they are not submodules of `(E →ₗ[R] F)`. I'll try to prove the equiv using my original definition instead.